File: flutter_survey_intercept/android/src/main/kotlin/com/example/flutter_survey_intercept/FlutterSurveyIntercept.kt
package com.example.flutter_survey_intercept
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.questionpro.cxlib.QuestionProCX
import com.questionpro.cxlib.enums.DataCenter
import com.questionpro.cxlib.interfaces.IQuestionProInitCallback
import com.questionpro.cxlib.interfaces.IQuestionProCallback
import com.questionpro.cxlib.model.TouchPoint
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class FlutterSurveyIntercept : FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var cxChannel: MethodChannel
    private var applicationContext: Context? = null
    private var isInitialized = false
    private var hasRepliedToInit = false
    private val mainHandler = Handler(Looper.getMainLooper())

    companion object {
        private const val CHANNEL = "intercept_sdk"
        private const val CX_CHANNEL = "Cx_Callback"
        private const val TAG = "FlutterSurveyIntercept"
        private const val API_KEY_META_NAME = "cx_manifest_api_key"
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext

        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)

        cxChannel = MethodChannel(binding.binaryMessenger, CX_CHANNEL)
        cxChannel.setMethodCallHandler { call, result ->
            if (call.method == "setScreenVisited") {
                setScreenVisited(call.argument("screen_name_key"), result)
            } else {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        cxChannel.setMethodCallHandler(null)
        applicationContext = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> initialize(call.argument("dataCenter") ?: "US", result)
            "setDataMappings" -> setDataMappings(call.argument("customVariables"), result)
            else -> result.notImplemented()
        }
    }

    private fun getApiKeyFromManifest(context: Context): String? {
        return try {
            val appInfo = context.packageManager.getApplicationInfo(
                context.packageName,
                android.content.pm.PackageManager.GET_META_DATA
            )
            appInfo.metaData?.getString(API_KEY_META_NAME)
        } catch (e: Exception) {
            Log.e(TAG, "Error reading API key from manifest", e)
            null
        }
    }

    private fun parseDataCenter(dataCenterStr: String): DataCenter {
        return try {
            DataCenter.valueOf(dataCenterStr.uppercase())
        } catch (e: IllegalArgumentException) {
            DataCenter.US
        }
    }

    private fun initialize(dataCenterStr: String, result: MethodChannel.Result) {
        val ctx = applicationContext ?: run {
            result.error("NO_CONTEXT", "Application context not available", null)
            return
        }

        if (isInitialized) {
            result.success("SDK already initialized")
            return
        }

        val apiKey = getApiKeyFromManifest(ctx)
        if (apiKey.isNullOrEmpty()) {
            result.error("MISSING_API_KEY", "API key not found in AndroidManifest.xml", null)
            return
        }

        try {
            val dataCenter = parseDataCenter(dataCenterStr)
            val touchPoint = TouchPoint.Builder(dataCenter).build()

            hasRepliedToInit = false

            QuestionProCX.getInstance().init(ctx, touchPoint, object : IQuestionProInitCallback {
                override fun onInitializationSuccess(message: String?) {
                    synchronized(this@FlutterSurveyIntercept) {
                        if (!hasRepliedToInit) {
                            hasRepliedToInit = true
                            isInitialized = true
                            result.success(message ?: "SDK initialized successfully")
                            getSurveyUrl()
                        }
                    }
                }

                override fun onInitializationFailure(error: String?) {
                    synchronized(this@FlutterSurveyIntercept) {
                        if (!hasRepliedToInit) {
                            hasRepliedToInit = true
                            result.error("INIT_FAILED", error ?: "Initialization failed", null)
                        }
                    }
                }
            })
        } catch (e: Exception) {
            Log.e(TAG, "Exception during SDK initialization", e)
            if (!hasRepliedToInit) {
                hasRepliedToInit = true
                result.error("INIT_EXCEPTION", e.message ?: "Unknown error occurred", null)
            }
        }
    }

    private fun setScreenVisited(screenName: String?, result: MethodChannel.Result) {
        if (screenName.isNullOrEmpty()) {
            result.error("INVALID_ARGS", "screen_name_key is required", null)
            return
        }

        try {
            QuestionProCX.getInstance().setScreenVisited(screenName)
            result.success("Event logged")
            getSurveyUrl()
        } catch (e: Exception) {
            result.error("SCREEN_VIEW_ERROR", e.message ?: "Unknown error occurred", null)
        }
    }

    private fun setDataMappings(customVariables: Map<String, String>?, result: MethodChannel.Result) {
        if (customVariables == null || customVariables.isEmpty()) {
            result.error("INVALID_ARGS", "customVariables cannot be null or empty", null)
            return
        }

        try {
            val customVars = HashMap<String, String>(customVariables)
            QuestionProCX.getInstance().setDataMappings(customVars)
            result.success("Data mappings set successfully")
            getSurveyUrl()
        } catch (e: Exception) {
            result.error("DATA_MAPPING_ERROR", e.message, null)
        }
    }

    /**
     * Callback listener triggered automatically when SDK has a survey URL available
     */
    private fun getSurveyUrl() {
        try {
            QuestionProCX.getInstance().getSurveyUrl(object : IQuestionProCallback {
                override fun getSurveyUrl(surveyUrl: String?) {
                    surveyUrl?.takeIf { it.isNotEmpty() }?.let { url ->
                        mainHandler.post {
                            channel.invokeMethod("onSurveyUrlReceived", mapOf("surveyUrl" to url))
                        }
                    }
                }
            })
        } catch (e: Exception) {
            Log.e(TAG, "Error setting up survey URL callback", e)
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {}
    override fun onDetachedFromActivity() {}
    override fun onDetachedFromActivityForConfigChanges() {}
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}
}

File: flutter_survey_intercept/ios/Classes/FlutterSurveyIntercept.swift
import Flutter
import UIKit
import QuestionProCXFramework

public class FlutterSurveyIntercept: NSObject, FlutterPlugin, QuestionProInitDelegate, QuestionProCallbackDelegate {

    private var window: UIWindow?
    private var isInitialized = false
    private var initResult: FlutterResult?
    private var surveyUrlChannel: FlutterMethodChannel?
    private var hasRepliedToInit = false
    private let initLock = NSLock()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "intercept_sdk", binaryMessenger: registrar.messenger())
        let cxChannel = FlutterMethodChannel(name: "Cx_Callback", binaryMessenger: registrar.messenger())
        let instance = FlutterSurveyIntercept()
        instance.surveyUrlChannel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addMethodCallDelegate(instance, channel: cxChannel)

        if let appDelegate = UIApplication.shared.delegate as? FlutterAppDelegate {
            instance.window = appDelegate.window
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        
        switch call.method {
        case "initialize":
            initialize(args: args, result: result)
        case "setScreenVisited":
            setScreenVisited(args: args, result: result)
        case "setDataMappings":
            handleSetDataMappings(args: args, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Method Handlers
    
    private func initialize(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let apiKey = args?["apiKey"] as? String, !apiKey.isEmpty else {
            result(FlutterError(code: "INVALID_ARGS", message: "apiKey is required for iOS", details: nil))
            return
        }
        
        let dataCenter = args?["dataCenter"] as? String ?? "US"
        initializeSurvey(apiKey: apiKey, dataCenter: dataCenter, result: result)
    }
    
    private func setScreenVisited(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let screenName = args?["screen_name_key"] as? String, !screenName.isEmpty else {
            result(FlutterError(code: "INVALID_ARGS", message: "screen_name_key is required", details: nil))
            return
        }
        
        setScreenName(screenName: screenName, result: result)
    }
    
    private func handleSetDataMappings(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let customVariables = args?["customVariables"] as? [String: String], !customVariables.isEmpty else {
            result(FlutterError(code: "INVALID_ARGS", message: "customVariables cannot be null or empty", details: nil))
            return
        }
        
        setDataMappings(customVariables: customVariables, result: result)
    }

    // MARK: - SDK Initialization

    private func initializeSurvey(apiKey: String, dataCenter: String, result: @escaping FlutterResult) {
        if isInitialized {
            result("SDK already initialized")
            return
        }

        guard let appWindow = window ?? UIApplication.shared
            .connectedScenes
            .compactMap({ ($0 as? UIWindowScene)?.windows.first })
            .first else {
            result(FlutterError(code: "NO_WINDOW", message: "Window not available", details: nil))
            return
        }

        let dataCenterValue = dataCenter.uppercased() == "EU" 
            ? TouchPoint.DataCenter.DATA_CENTER_EU 
            : TouchPoint.DataCenter.DATA_CENTER_US

        let touchPoint = TouchPoint.initTouchPoint(dataCenter: dataCenterValue)
        touchPoint.platform = TouchPoint.PLATFORM.FLUTTER.rawValue
        
        initResult = result
        initLock.lock()
        hasRepliedToInit = false
        initLock.unlock()

        do {
            QuestionProCX.getinstance().configure(
                apiKey: apiKey,
                touchPoint: touchPoint,
                withWindow: appWindow,
                initCallbackDelegate: self,      
            )
            QuestionProCX.getinstance().setQuestionProCallbackDelegate(questionProCallbackDelegate: self)
        } catch {
            initLock.lock()
            defer { initLock.unlock() }
            if !hasRepliedToInit {
                hasRepliedToInit = true
                initResult?(FlutterError(
                    code: "INIT_EXCEPTION",
                    message: "SDK initialization failed: \(error.localizedDescription)",
                    details: nil
                ))
                initResult = nil
            }
        }
    }

    // MARK: - Screen View & Data Mappings

    private func setScreenName(screenName: String, result: @escaping FlutterResult) {
        do {
            QuestionProCX.getinstance().setScreenName(screenName: screenName)
            result("Event logged")
        } catch {
            result(FlutterError(code: "SCREEN_VIEW_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func setDataMappings(customVariables: [String: String], result: @escaping FlutterResult) {
        do {
            QuestionProCX.getinstance().setDataMappings(dataMappings: customVariables)
            result("Data mappings set successfully")
        } catch {
            result(FlutterError(code: "DATA_MAPPING_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    // MARK: - QuestionPro Delegate Callbacks

    public func initSDKSuccess() {
        initLock.lock()
        defer { initLock.unlock() }
        if !hasRepliedToInit {
            hasRepliedToInit = true
            isInitialized = true
            initResult?("QuestionPro CX SDK initialized successfully")
            initResult = nil
        }
    }

    public func initSDKFailed(error: String) {
        initLock.lock()
        defer { initLock.unlock() }
        if !hasRepliedToInit {
            hasRepliedToInit = true
            initResult?(FlutterError(code: "INIT_FAILED", message: error.isEmpty ? "Initialization failed" : error, details: nil))
            initResult = nil
        }
    }
    
    public func getSurveyURL(surveyURL: String) {
        guard !surveyURL.isEmpty else { return }
        surveyUrlChannel?.invokeMethod("onSurveyUrlReceived", arguments: ["surveyUrl": surveyURL])
    }
    
    deinit {
        initResult = nil
        window = nil
        surveyUrlChannel = nil
    }
}
